File: go/src/interpreter/interpreter.go
package interpreter

import (
	"errors"
	"finamore00/brainfrick/src/programstack"
	"fmt"
	"io"
	"os"
	"slices"
)

type Interpreter struct {
	src    *os.File
	input  io.ByteReader
	memory []byte
	memCap uint64
	memPtr uint64
	pc     uint64
	st     programstack.Stack[uint64]
}

func New(src *os.File, input io.ByteReader, memCap uint64) *Interpreter {
	return &Interpreter{
		src:    src,
		input:  input,
		memory: make([]byte, memCap),
		memCap: memCap,
		pc:     0,
		memPtr: 0,
		st:     programstack.Stack[uint64]{},
	}
}

func (i *Interpreter) Run() error {
	buff := make([]byte, 1) //Stupid solution to be able to read a file byte by byte

	for _, err := i.src.Read(buff); err != io.EOF; _, err = i.src.Read(buff) {
		switch buff[0] {
		case '+':
			i.memory[i.memPtr] += 1
		case '-':
			i.memory[i.memPtr] -= 1
		case '<':
			if i.memPtr > 0 {
				i.memPtr -= 1
			}
		case '>':
			if i.memPtr < i.memCap-1 {
				i.memPtr += 1
			}
		case '.':
			fmt.Printf("%c", i.memory[i.memPtr])
		case ',':
			i.memory[i.memPtr], err = i.input.ReadByte()
			if err != nil {
				return err
			}
		case '[':
			if i.memory[i.memPtr] == 0 {
				tokens := []byte{'+', '-', '<', '>', '.', ',', '[', ']'}

				//Look for matching ']'
				depth := uint64(1)
				for _, err = i.src.Read(buff); depth > 0 && err != io.EOF; _, err = i.src.Read(buff) {
					if buff[0] == '[' {
						depth += 1
					}

					if buff[0] == ']' {
						depth -= 1
					}

					if slices.Contains(tokens, buff[0]) {
						i.pc += 1
					}
				}

				if err == io.EOF {
					return errors.New("ERROR: Found '[' token with no matching ']'")
				}
			} else {
				i.st.Push(i.pc)
			}
		case ']':
			if i.st.IsEmpty() {
				return errors.New("ERROR: Found ']' token with no matching '['")
			}

			if i.memory[i.memPtr] != 0 {
				tokens := []byte{'+', '-', '<', '>', '.', ',', '[', ']'}

				tgtPc, _ := i.st.Peek()
				for i.pc > tgtPc {
					i.src.Seek(-2, io.SeekCurrent)
					i.src.Read(buff)

					if slices.Contains(tokens, buff[0]) {
						i.pc -= 1
					}
				}
			} else {
				i.st.Pop()
			}
		default:
			continue //Everything else is dismissed as comment
		}

		i.pc += 1
	}
	return nil
}

File: c-sharp/src/Interpreter/Interpreter.cs
namespace brainfrick.interpreter;

using System.IO;

public class InvalidSyntaxException(string message) : Exception(message);

public class Interpreter(FileStream src, ulong cap)
{
    private readonly FileStream src = src;
    private readonly byte[] memory = new byte[cap];
    private readonly ulong MemCap = cap;
    private ulong MemPtr = 0;
    private ulong pc = 0;
    private Stack<ulong> st = new();

    public void Run()
    {
        char[] tokens = ['+', '-', '>', '<', ',', '.', '[', ']'];

        for (int c = src.ReadByte(); c != -1; c = src.ReadByte())
        {
            switch ((char)c)
            {
                case '+':
                    memory[MemPtr] += 1;
                    break;
                case '-':
                    memory[MemPtr] -= 1;
                    break;
                case '>':
                    MemPtr = MemPtr < MemCap + 1 ? MemPtr + 1 : MemPtr;
                    break;
                case '<':
                    MemPtr = MemPtr > 0 ? MemPtr - 1 : MemPtr;
                    break;
                case '.':
                    Console.Write((char)memory[MemPtr]);
                    break;
                case ',':
                    memory[MemPtr] = (byte)Console.ReadKey().KeyChar;
                    break;
                case '[':

                    if (memory[MemPtr] == 0)
                    {
                        //Look for matching ']'
                        ulong depth = 1;
                        for (c = src.ReadByte(); depth > 0 && c != -1; c = src.ReadByte())
                        {
                            if ((char)c == '[') depth += 1;
                            if ((char)c == ']') depth -= 1;

                            if (tokens.Contains((char)c)) pc++;
                        }

                        if (c == -1)
                        {
                            throw new InvalidSyntaxException("ERROR: Found '[' token without matching ']'.");
                        }
                    }
                    else
                    {
                        st.Push(pc);
                    }
                    break;
                case ']':
                    if (st.Count <= 0)
                    {
                        throw new InvalidSyntaxException("ERROR: Found ']' without matching '['");
                    }

                    if (memory[MemPtr] != 0)
                    {
                        ulong TgtPc = st.Peek();
                        while (pc > TgtPc)
                        {
                            src.Seek(-2, SeekOrigin.Current);
                            if (tokens.Contains((char)src.ReadByte())) pc--;
                        }
                    }
                    else
                    {
                        st.Pop();
                    }
                    break;
                default:
                    continue; //Everything else is dismissed as a comment
            }

            pc++;
        }
    }
}